Subbus/Classes/Subbus.Scoping.swift

```diff
@@ -0,0 +1,75 @@
+//
+//  Subbus.Scoping.swift
+//  Pods-Subbus_Example
+//
+//  Created by Alex Larson on 2/6/20.
+//
+
+/// The protocol defining the scoping capabilities added to Subbus.
+protocol SubbusScopingProtocol {
+    /// Adds a new subscription to Subbus with a `ScopedEvent`.
+    ///
+    /// This scope limits the receivers of an event even further,
+    /// allowing the same event to have further separation of types without defining multiple events.
+    ///
+    /// There is no easy way to unsubscribe from a particular scope by using ONLY the scope.
+    /// I.e. there is no `unsubscribe(id:event:scope:)`.
+    ///
+    /// If you want to unsubscribe from with a particular scope - use a unique ID (for example - reuse the scope!).
+    ///
+    ///     let scope = "abc"
+    ///     addSubscription(id: scope, MyEvent.self, limitedToScope: scope)
+    ///     unsubscribe(id: scope, MyEvent.self)
+    ///
+    /// Parameters:
+    ///     - id: The id of the subscriber to this event.  Used for unregistering.
+    ///     - event: The event type to register the subscriber to.  This class will be a subclass of `ScopedEvent`.
+    ///     - limitedToScope: The scope to subscribe to.  Cannot be `nil` or an `Optional` type.
+    ///     - callback: The handler block for when the event is `Post`ed.
+    static func addSubscription<S, I, T>(id: I, event: T.Type, limitedToScope scope: S, callback: @escaping (T) -> Void)
+    
+    /// Posts a `PersistentEvent` to the subscribed receivers.
+    ///
+    /// Parameters:
+    ///     - event: The event to be posted.
+    ///     - limitedToScope: The scope limitation for the posted event.  Cannot be `nil` or an `Optional` type.
+    static func post<S, T>(event: T, limitedToScope scope: S)
+}
+
+/// Conformance to the `SubbusScopingProtocol`.
+extension Subbus: SubbusScopingProtocol {
+    static func addSubscription<S, I, T>(id: I, event: T.Type, limitedToScope scope: S, callback: @escaping (T) -> Void) {
+        //Verify data
+        guard (scope as? OptionalProtocol) == nil else { log("Subscribe - You may not pass in an optional scope."); return }
```

Comment:
Maybe we should change this to throwing an exception? I feel like it could be easy for somebody to attempt an optional subscription, overlooking the log, and spending time trying to debug there app.

Reply:
Yeah, I can see that happening.  And in fact, since `force:` isn't passed, it actually wouldn't be logged unless they set `Subbus.logToConsole = true`.

I'd like to avoid Subbus making apps crash, I'd rather it blew up the console log than crashed the app.  Maybe we just make it crash on debug builds?

Reply:
I've made a workaround:

```
    private static func logError(_ error: String) {
        #if DEBUG
        fatalError(error)
        #else
        log(error, force: true)
        #endif
    }
```

Reply:
I'm inclined to agree, but we aren't making the decision to crash the app. The app that implements the library should be responsible for implementing a try-catch (or a simple `try?`)

I do hate try-catch blocks, but I worry that we are being _too_ nice right now.